Update and add dev guide content

rhornung67 opened the pull request:
# Summary

- This PR adds content to the RAJA Dev Guide about CI test organization and maintenance tasks.

The generated documentation can be viewed here: https://raja.readthedocs.io/en/task-ci-docs/sphinx/dev_guide/index.html

kab163 commented:
Is there a link to the readthedocs that this generates?

kab163 reviewed (commented):
Should there be any mentioning of how to create a new spec job? (Or updating a job if updates are done to a LC machine to where a spec no longer works?) That may require adding certain flags in the `scripts/spack_packages/raja/package.py` file or adding other parameters to the spec like `cxxflags=...`

rhornung67 commented:
> Should there be any mentioning of how to create a new spec job? (Or updating a job if updates are done to a LC machine to where a spec no longer works?) That may require adding certain flags in the `scripts/spack_packages/raja/package.py` file or adding other parameters to the spec like `cxxflags=...`

Yes. I am planning to add a section on CI maintenance tasks that will cover those sorts of things.

rhornung67 commented on docs/sphinx/dev_guide/build_configurations.rst:
```diff
@@ -113,18 +117,18 @@ in sub-directories in the RAJA ``scripts/uberenv/spack_configs`` directory:
   config.yaml
```
that section is old and needs to be updated...

rhornung67 replied:
fixed.

rhornung67 commented on docs/sphinx/dev_guide/build_configurations.rst:
```diff
@@ -135,8 +139,7 @@ the compiler and version. For example,
 Specs that are exercised during the Gitlab CI process are found YAML files in 
-the ``RAJA/.gitlab`` directory. See :ref:`vettedspecs-label` for more 
-information.
+the ``RAJA/.gitlab`` directory. See :ref:`ci-label` for more information.
```
ditto.

rhornung67 replied:
fixed.

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,78 +12,329 @@
+
+.. figure:: ./figures/RAJA-Gitlab-Files.png
+
+   Directories and files in the RAJA repo that support Gitlab CI testing.
```
That's the figure caption.

kab163 replied:
Perhaps "This figure shows the directories and files..."

rhornung67 replied:
fixed.

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,78 +12,329 @@
+In the ``stages`` section above, prefixes 'r_', 'l_', and 'c_' refer to 
+resources in the LC on which tests are run. Specifically, the machines 'ruby',
+'lassen' and 'corona'. The jobs that will run in the pipeline(s) for each 
+resource are defined in the files listed in the ``include`` section. 
```
Dunno. I must have been sleepy.... I will fix.

kab163 commented:
I will think about this over the weekend. If I think about other things that could be added here, I'll let you know. So far, this looks like very helpful, much needed documentation. Thanks Rich!

rhornung67 commented on docs/sphinx/dev_guide/build_configurations.rst:
```diff
@@ -97,11 +100,12 @@ Generating a RAJA host-config file
 
 Compiler configurations for Livermore computer platforms are contained in 
-in sub-directories in the RAJA ``scripts/uberenv/spack_configs`` directory:
+in sub-directories in the ``RAJA/scripts/uberenv/spack_configs`` directory.
```
fixed.

rhornung67 commented on docs/sphinx/dev_guide/build_configurations.rst:
```diff
@@ -164,29 +167,34 @@ MacOS
+corresponding to the OS version in the ``compilers.yaml`` file in the 
+``RAJA/scripts/radiuss-spack-configs/darwin/`` directory, and a commented 
+section to illustrate how to add `CMake` as an external package in the
+``packages.taml`` in the same directory. You may also install CMake 
```
fixed.

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,78 +12,329 @@
+The figure below shows the high-level steps in the RAJA Gitlab CI testing 
+process. The main steps, which we will discuss in more detail later, are:
+
+  #. *Mirror* the RAJA GitHub repo to the RAJA project in the Gitlab instance. 
```
I address your first concern. 

I'm not clear on what to say about mirroring. If you have any thoughts to share about what someone unfamiliar with it should know, please let me know, or feel free to edit the docs.

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,78 +12,329 @@
+
+In the ``stages`` section above, prefixes 'r_', 'l_', and 'c_' refer to 
+resources in the LC on which tests are run. Specifically, the machines 'ruby',
+'lassen' and 'corona'. The jobs that will run in the pipeline(s) for each 
```
done.

rhornung67 commented on docs/sphinx/dev_guide/ci.rst:
```diff
@@ -12,78 +12,329 @@
+'lassen' and 'corona'. The jobs that will run in the pipeline(s) for each 
+resource are defined in the files listed in the ``include`` section. 
+
+The ``RAJA/.gitlab`` directory contains a *resource* and *jobs* file for each 
```
fixed.

rhornung67 commented:
@kab163 I addressed all your comments, except for adding more info about Gitlab mirroring. I also, added links to files in the repo that may help folks navigate more easily.

rhornung67 commented:
@mdavis36 if you have some time, please take a crack at filling in some content in the Azure CI maintenance tasks section in the dev guide, specifically, w.r.t. to the docker images we use for the builds.

MrBurmark commented:
This looks good. I learning a lot about the CI config and spack stuff. I also liked the section about writing a test.

rhornung67 commented:
Thanks @MrBurmark. I will take another clean up pass and add a few missing things that I recently learned. Then, I will remove draft status and it should be good to go.

rhornung67 commented:
Thanks for the approvals. Anyone else want to review and approve before I merge?

rchen20 commented:
Taking a quick look now.

rchen20 commented on docs/sphinx/dev_guide/ci_tasks.rst:
```diff
@@ -0,0 +1,234 @@
+
+      . /opt/spack/share/spack/setup-env.sh && spack load hip llvm-amdgpu
+
+    **Sycl** requiers us to run setupvars.sh::
```
```suggestion
    **Sycl** requires us to run setupvars.sh::
```

rhornung67 replied:
fixed.

rchen20 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -24,19 +31,19 @@ The RAJA release process typically involves the following sequence of steps:
      into the **main branch.** When it is approved and all CI checks pass,
```
Just to clarify, we create a release candidate branch from `develop`, but merge it into `main`?

rchen20 replied:
Thanks, so `main` only has updates up until the latest release.

rhornung67 replied:
that is correct.

rchen20 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -134,3 +146,36 @@ the key steps for performing a hotfix release are:
+    updating the submodules to the the new RAJA Portability Suite project 
+    versions, making sure the proxy-apps build and run correctly. When this
+    is done, tag a release for proxy-app project.
+  * Update the `RAJAProxies <https://github.com/LLNL/RAJAProxies>`_ project
```
RAJAProxies mentioned twice, although both statements are slightly different.

rchen20 commented on docs/sphinx/dev_guide/release_process.rst:
```diff
@@ -134,3 +146,32 @@ the key steps for performing a hotfix release are:
+
+  * Update the `RAJAProxies <https://github.com/LLNL/RAJAProxies>`_ project
+    to the newly RAJA Portability Suite projects. This typically consists of 
+    updating the submodules to the the new RAJA Portability Suite project 
```
```suggestion
    updating the submodules to the new RAJA Portability Suite project 
```

rhornung67 commented:
Folks, I took a last pass and fixed several typos, etc. unfortunately, this dismissed your approvals. Please re-approve and I'll merge this when it comes back green.
